api/grpcserver/v2alpha1/transaction.go

```diff
@@ -0,0 +1,463 @@
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Failed to verify transaction: %s", err.Error()))
+	}
+
+	if err := s.publisher.Publish(ctx, pubsub.TxProtocol, request.Transaction); err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to publish transaction: %s", err.Error()))
+	}
+
+	raw := types.NewRawTx(request.Transaction)
+	return &spacemeshv2alpha1.SubmitTransactionResponse{
+		Status: &rpcstatus.Status{Code: int32(code.Code_OK)},
+		TxId:   raw.ID[:],
+	}, nil
+}
+
+func (s *TransactionService) EstimateGas(
+	ctx context.Context,
+	request *spacemeshv2alpha1.EstimateGasRequest,
+) (*spacemeshv2alpha1.EstimateGasResponse, error) {
+	if len(request.Transaction) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "transaction is empty")
+	}
+	raw := types.NewRawTx(request.Transaction)
+	req := s.conState.Validation(raw)
+	// TODO: Fill signature if it's not present
+	header, err := req.Parse()
+	if errors.Is(err, core.ErrNotSpawned) {
+		return nil, status.Error(codes.NotFound, "account is not spawned")
+	} else if errors.Is(err, core.ErrMalformed) {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	} else if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &spacemeshv2alpha1.EstimateGasResponse{
+		Status:            nil,
+		RecommendedMaxGas: header.MaxGas,
+	}, nil
+}
+
+func toTransactionOperations(filter *spacemeshv2alpha1.TransactionRequest) (builder.Operations, error) {
+	ops := builder.Operations{}
+	if filter == nil {
+		return ops, nil
+	}
+
+	if filter.GetPrincipal() != "" {
+		addr, err := types.StringToAddress(filter.GetPrincipal())
+		if err != nil {
+			return builder.Operations{}, err
+		}
+		ops.Filter = append(ops.Filter, builder.Op{
+			Field: builder.Principal,
+			Token: builder.Eq,
+			Value: addr.Bytes(),
+		})
+	}
+
+	if len(filter.Txid) > 0 {
+		ops.Filter = append(ops.Filter, builder.Op{
+			Field: builder.Id,
+			Token: builder.In,
+			Value: filter.Txid,
+		})
+	}
+
+	if filter.StartLayer != nil {
+		ops.Filter = append(ops.Filter, builder.Op{
+			Field: builder.Layer,
+			Token: builder.Gte,
+			Value: int64(filter.GetStartLayer()),
+		})
+	}
+
+	if filter.EndLayer != nil {
+		ops.Filter = append(ops.Filter, builder.Op{
+			Field: builder.Address,
+			Token: builder.Lte,
+			Value: int64(filter.GetEndLayer()),
+		})
+	}
+
+	if len(ops.Filter) > 0 {
+		ops.StartWith = "and"
+	}
+
+	ops.Modifiers = append(ops.Modifiers, builder.Modifier{
+		Key:   builder.OrderBy,
+		Value: "layer asc, id",
+	})
+
+	if filter.Limit != 0 {
+		ops.Modifiers = append(ops.Modifiers, builder.Modifier{
+			Key:   builder.Limit,
+			Value: int64(filter.Limit),
+		})
+	}
+	if filter.Offset != 0 {
+		ops.Modifiers = append(ops.Modifiers, builder.Modifier{
+			Key:   builder.Offset,
+			Value: int64(filter.Offset),
+		})
+	}
+	return ops, nil
+}
+
+func (s *TransactionService) toTx(tx *types.MeshTransaction, result *types.TransactionResult,
```

Comment:
why is this a method of `TransactionService` but ` toTransactionOperations` is not? Both don't access fields of the service.

Reply:
That was leftover from previous idea :) Fixed.